src/webgpu/shader/execution/expression/call/builtin/subgroupMul.spec.ts

```diff
@@ -0,0 +1,374 @@
+  numberToFloatBits,
+  floatBitsToNumber,
+  kFloat32Format,
+  kFloat16Format,
+  scalarTypeOf,
+} from '../../../../../util/conversion.js';
+import { FP } from '../../../../../util/floating_point.js';
+
+import {
+  kNumCases,
+  kStride,
+  kWGSizes,
+  kPredicateCases,
+  runAccuracyTest,
+  runComputeTest,
+} from './subgroup_util.js';
+
+export const g = makeTestGroup(GPUTest);
+
+const kIdentity = 1;
+
+const kDataTypes = objectsToRecord(kConcreteNumericScalarsAndVectors);
+
+const kOperations = ['subgroupMul', 'subgroupExclusiveMul'] as const;
+
+g.test('fp_accuracy')
+  .desc(
+    `Tests the accuracy of floating-point multiplication.
+
+The order of operations is implementation defined, most threads are filled with
+the identity value and two receive random values.
+Subgroup sizes are not known ahead of time so some cases may not perform any
+interesting operations. The test biases towards checking subgroup sizes under 64.
+These tests only check two values in order to reuse more of the existing infrastructure
+and limit the number of permutations needed to calculate the final result.`
+  )
+  .params(u =>
+    u
+      .combine('case', [...iterRange(kNumCases, x => x)])
+      .combine('type', ['f32', 'f16'] as const)
+      .combine('wgSize', [
+        [kStride, 1, 1],
+        [kStride / 2, 2, 1],
+      ] as const)
+  )
+  .beforeAllSubcases(t => {
+    const features: GPUFeatureName[] = ['subgroups' as GPUFeatureName];
+    if (t.params.type === 'f16') {
+      features.push('shader-f16');
+      features.push('subgroups-f16' as GPUFeatureName);
+    }
+    t.selectDeviceOrSkipTestCase(features);
+  })
+  .fn(async t => {
+    await runAccuracyTest(
+      t,
+      t.params.case,
+      [t.params.wgSize[0], t.params.wgSize[1], t.params.wgSize[2]],
+      'subgroupMul',
+      t.params.type,
+      kIdentity,
+      t.params.type === 'f16' ? FP.f16.multiplicationInterval : FP.f32.multiplicationInterval
+    );
+  });
+
+/**
+ * Checks subgroup multiplications.
+ *
+ * Expected results:
+ * - subgroupMul: each invocation should have result equal to real subgroup size
+ * - subgroupExclusiveMul: each invocation should have result equal to its subgroup invocation id
+ * @param metadata An array containing actual subgroup size per invocation followed by
+ *                 subgroup invocation id per invocation
+ * @param output An array of multiplications
+ * @param type The data type
+ * @param operation Type of multiplication
+ */
+function checkMultiplication(
+  metadata: Uint32Array,
+  output: Uint32Array,
+  type: Type,
+  operation: 'subgroupMul' | 'subgroupExclusiveMul'
+): undefined | Error {
+  let numEles = 1;
+  if (type instanceof VectorType) {
+    numEles = type.width;
+  }
+  const scalarTy = scalarTypeOf(type);
+  const expectedOffset = operation === 'subgroupMul' ? 0 : metadata.length / 2;
+  for (let i = 0; i < metadata.length / 2; i++) {
+    const expected = Math.pow(2, metadata[i + expectedOffset]);
```

Comment:
Ah, this is used with limited subgroup sizes.

I had this long comment where for the float cases I used fill value -1.0625 = -(1 + ldexp(1,-4)).... which works nicely but very much complicates the predicate.

Never mind, SGTM.